Revise @Poller usage to single entry in Messaging Annotations.

onobc opened the pull request:
Fixes #3506

onobc committed:
Revise @Poller usage to single entry in Messaging Annotations.

Fixes #3506

onobc committed:
Add @Author

artembilan commented on src/reference/asciidoc/changes-5.5-6.0.adoc:
```diff
@@ -0,0 +1,26 @@
+=== General Changes
+
+[[x6.0-poller]]
+==== @Poller
```
This info has to be added to the `whats-new.adoc`. Because this is exactly what is new in this `6.0` version.
We will create a `changes-5.5-6.0.adoc` file, when we start `6.1` version.

The point is to let end-users to find easily what is new in the project just from a first page of our doc.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/MessagingAnnotationUtils.java:
```diff
@@ -69,9 +70,31 @@
+			return false;
+		}
+		// Poller with special 'none' value
+		if ((annotationValue instanceof Poller) && ValueConstants.DEFAULT_NONE.equals(((Poller) annotationValue).value())) {
```
This is not OK. 
Better really to have an utility method as generics as possible.
I mean I believe there is a way to get a value from the annotation without casting to some specific type.
For example see the same `@Reactive` use-case:
```
Reactive reactive() default @Reactive(ValueConstants.DEFAULT_NONE);
```
It is fully similar to what we have so far with the `@Poller`.


Also: if we do this then we don't need that extra check `if (poller != null && !ValueConstants.DEFAULT_NONE.equals(poller.value())) {` in the `AbstractMethodAnnotationPostProcessor.java`

onobc replied:
Yep, I meant to add a comment on this controversial section. The alternative approach I was going to talk about would be to consider any annotation whose "value" field was set to `ValueConstants.DEFAULT_NONE` to be considered "empty". This would take the `Poller` or `Reactive` type out of the equation. This would also handle `@Reactive` annotation extension properly. 

> For example see the same @Reactive use-case: 
> It is fully similar to what we have so far with the @Poller.

AFAICT we do not handle `@Reactive` override in annotation hierarchy properly (you will see it is not in the EnableIntegrationTests). We would need to add the special handling for `@Reactive` as well. 

I will amend the proposal w/ the above approach and we can see how that looks.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/MessagingAnnotationUtils.java:
```diff
@@ -69,9 +70,31 @@
-		return value != null && (!(value instanceof String) || (StringUtils.hasText((String) value))) // NOSONAR
-				&& (!value.getClass().isArray() || ((Object[]) value).length > 0);
+	/**
+	 * Determines if the value of a named attribute from an annotation instance contains an actual value.
```
The method JavaDoc must be imperative, kinda commanding us what to do.
No blank lines in the JavaDoc: https://github.com/spring-projects/spring-framework/wiki/Code-Style#javadoc-formatting

artembilan reviewed (changes requested):
Thanks, Chris.

Please, find some review on the lines.

onobc committed:
Code review feedback

- Make annotation utils check for NONE more generic
- Move docs to whats-new.adoc
- Polish javadoc

artembilan commented on src/reference/asciidoc/whats-new.adoc:
```diff
@@ -36,4 +36,21 @@ See <<./http.adoc#http,HTTP Support>> for more information.
+- `@Splitter`
+- `@Transformer`
+
+used to contain a `Poller[]` attribute but now contain a single element. The array was only in place  to allow an empty default (no poller). This is now accomplished with a single element using the special value of `ValueConstants.DEFAULT_NONE`.
```
Hm. I thought I send you a suggestion how to make it clean and simple:

> The messaging annotations don't require a `poller` attribute as an array of `@Poller` any more.
See <<./configuration.adoc#configuration-using-poller-annotation,Using the `@Poller` Annotation>> for more information.

And I believe this one has to be as a plain sentence in the existing `=== General Changes` section: no need for its own since it is really a general one.

artembilan commented:
FYI: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-5.x-to-6.0-Migration-Guide#no-array-for-poller-annotation-attribute

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -363,24 +364,26 @@ protected AbstractEndpoint createEndpoint(MessageHandler handler, @SuppressWarni
 
-		Poller[] pollers = MessagingAnnotationUtils.resolveAttribute(annotations, "poller", Poller[].class);
+		Poller poller = MessagingAnnotationUtils.resolveAttribute(annotations, "poller", Poller.class);
+		boolean pollerProvided = poller != null && !ValueConstants.DEFAULT_NONE.equals(poller.value());
```
Since the `ValueConstants.DEFAULT_NONE` is covered now in the `MessagingAnnotationUtils.hasValue()`, I guess we don't need it here any more.

onobc replied:
It is used [here](https://github.com/spring-projects/spring-integration/pull/3746/files#diff-4968e3d7ef5068b890a520af47f1fc8da4b9c369f7af56fbb05efbd5d07f070bR373) to ensure XOR on reactive|poller though.

onobc replied:
Yeh it is now simplified. Good code review @artembilan !!

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -414,19 +417,15 @@ private ReactiveStreamsConsumer reactiveStreamsConsumer(MessageChannel channel,
-					"The 'poller' for an Annotation-based endpoint can have only one '@Poller'.");
-			Poller poller = pollers[0];
-
+		if (poller != null && !ValueConstants.DEFAULT_NONE.equals(poller.value())) {
```
DITTO

onobc committed:
Code review feedback
- simplify docs
- simplify isProvided checks

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/f3d3694836150312e198c05de06e755105af2039 after some commit message clean up.

Thank you, Chris, again!
